flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -219,6 +219,18 @@ public TaskStateSnapshot retrieveLocalState(long checkpointID) {
             snapshot = loadTaskStateSnapshot(checkpointID);
         }
 
+        // Even if local recovery is disabled, it is still necessary to load the TaskStateSnapshot
```

Comment:
nit: I think `localRecoveryConfig.isLocalRecoveryEnabled()` can be checked first, which can avoid acquiring for the lock.